xgcm/test/datasets.py

```diff
@@ -238,3 +238,155 @@ def periodic_2d(request):
+    )
+
+    dy_ne = xr.DataArray(
+        np.ones([len(xt), len(yt)]) * dy + 0.1, coords=[("xu", xu), ("yu", yu)]
+    )
+    dy_n = xr.DataArray(
+        np.ones([len(xt), len(yt)]) * dy + 0.2, coords=[("xt", xt), ("yu", yu)]
+    )
+    dy_e = xr.DataArray(
+        np.ones([len(xt), len(yt)]) * dy + 0.3, coords=[("xu", xu), ("yt", yt)]
+    )
+    dy_t = xr.DataArray(
+        np.ones([len(xt), len(yt)]) * dy + 0.4, coords=[("xt", xt), ("yt", yt)]
+    )
+
+    dz_t = xr.DataArray(
+        data_generator() * dz, coords=[("xt", xt), ("yt", yt), ("time", t), ("zt", zt)]
+    )
+    dz_w = xr.DataArray(
+        data_generator() * dz, coords=[("xt", xt), ("yt", yt), ("time", t), ("zw", zw)]
+    )
+
+    # Make sure the areas are not just the product of x and y distances
+    area_ne = (dx_ne * dy_ne) + 0.1
+    area_n = (dx_n * dy_n) + 0.2
+    area_e = (dx_e * dy_e) + 0.3
+    area_t = (dx_t * dy_t) + 0.4
+
+    # calculate volumes, but again add small differences.
+    volume_t = (dx_t * dy_t * dz_t) + 0.25
+
+    def _add_metrics(obj):
+        obj = obj.copy()
+        for name, data in [
+            ("dx_ne", dx_ne),
+            ("dx_n", dx_n),
+            ("dx_e", dx_e),
+            ("dx_t", dx_t),
+            ("dy_ne", dy_ne),
+            ("dy_n", dy_n),
+            ("dy_e", dy_e),
+            ("dy_t", dy_t),
+            ("dz_t", dz_t),
+            ("dz_w", dz_w),
+            ("area_ne", area_ne),
+            ("area_n", area_n),
+            ("area_e", area_e),
+            ("area_t", area_t),
+            ("volume_t", volume_t),
+        ]:
+            obj.coords[name] = data
+            obj.coords[name].attrs["tracked_name"] = name
+        # add xgcm attrs
+        for ii in ["xu", "xt"]:
+            obj[ii].attrs["axis"] = "X"
+        for ii in ["yu", "yt"]:
+            obj[ii].attrs["axis"] = "Y"
+        for ii in ["zt", "zw"]:
+            obj[ii].attrs["axis"] = "Z"
+        for ii in ["time"]:
+            obj[ii].attrs["axis"] = "T"
+        for ii in ["xu", "yu", "zw"]:
+            obj[ii].attrs["c_grid_axis_shift"] = 0.5
+        return obj
+
+    coords = {
+        "X": {"center": "xt", "right": "xu"},
+        "Y": {"center": "yt", "right": "yu"},
+        "Z": {"center": "zt", "right": "zw"},
+    }
+
+    metrics = {
+        ("X",): ["dx_t", "dx_n", "dx_e", "dx_ne"],
+        ("Y",): ["dy_t", "dy_n", "dy_e", "dy_ne"],
+        ("Z",): ["dz_t", "dz_w"],
+        ("X", "Y"): ["area_t", "area_n", "area_e", "area_ne"],
+        ("X", "Y", "Z"): ["volume_t"],
+    }
+
+    # combine to different grid configurations (B and C grid)
+    ds_b = _add_metrics(
+        xr.Dataset(
+            {"u": u_b, "v": v_b, "wt": wt, "tracer": tr, "timeseries": timeseries}
+        )
+    )
+    ds_c = _add_metrics(
+        xr.Dataset(
+            {"u": u_c, "v": v_c, "wt": wt, "tracer": tr, "timeseries": timeseries}
+        )
+    )
+
+    return {"B": ds_b, "C": ds_c, "coords": coords, "metrics": metrics}
```

Comment:
I would prefer to just return one dataset at a time.

Reply:
Could we do a separate PR for this? #132 rebased on this style and I think it is important to fix that bug. I promise Ill work swiftly on a PR that makes this dataset routine nicer. Then I can also add a non-uniform case.